Async opportunity search support (#240)
stapi-spec/stapi-spec: README.md

| `POST /products/{productId}/opportunities`               | Opportunities       | [Opportunity Request](./opportunity/README.md#opportunity-request) | Sync search: [Opportunities Collection](./opportunity/README.md#opportunities-collection); Async search: [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | Explore the opportunities available for a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters) |
| `GET /products/{productId}/opportunities/{collectionId}` | Async Opportunities | -                                                                  | [Opportunities Collection](./opportunity/README.md#opportunities-collection) | Get the opportunity collection for an async opportunity search              |

> | `GET /searches/opportunities/`                           | Async Opportunities | -                                                                  | [[Opportunity Search Record](./opportunity/README.md#opportunity-search-record)] |          |

Reviewer: I don't think this is an endpoint, maybe just a copy error?

> | `GET /searches/opportunities/`                           | Async Opportunities | -                                                                  | [[Opportunity Search Record](./opportunity/README.md#opportunity-search-record)] |          |

Reviewer: I'm unclear as to what this endpoint does. Why does it have a trailing slash? and why does it return the same type as /searches/opportunities/{searchRecordId}?

Author: It's hard to see in the raw form but this endpoint is returning an array of Opportunity Search Records. This endpoint allows a user to list all their searches in one place. The trailing slash was an error on my part and I will remove it.

root (`/`) document, while OGC API requires they also live at the
`/conformance` endpoint. STAPI's conformance structure is detailed in the
[core](core/). Note all conformance URIs serve up a rendered HTML version of
the corresponding OpenAPI document at the given location.

### Conformance Class Table

| **Name**                    | **Specified in**                       | **Conformance URI**                                  | **Description**                                                                          |
| --------------------------- | -------------------------------------- | ---------------------------------------------------- | ---------------------------------------------------------------------------------------- |
| STAPI - Core                | Core                                   | https://stapi.example.com/v0.1.0/core                | Specifies the STAPI Landing page `/`, communicating conformance and available endpoints. |
| STAPI - Opportunities       | [Opportunities](opportunity/README.md) | https://stapi.example.com/v0.1.0/opportunities       | Enables request of potential tasking opportunities                                       |

> | STAPI - Async Opportunities | [Opportunities](opportunity/README.md) | https://stapi.example.com/v0.1.0/async-opportunities | Enables async request of potential tasking opportunities                                 |

Reviewer: I'd prefer this to be `opportunities-async` so it's more apparently related to `opportunities`

| `GET /orders/{orderId}`                      | `order`            |
| `GET /orders/{orderId}/status`               | `monitor`          |

`create-order`: A link with this relation type should only be provided in the
landing page if a user can directly go from the products to the order endpoint
without going through the `POST /products/{productId}/opportunities` endpoint.

### Opportunities

The `POST /products/{productId}/opportunities` endpoint provides additional
functionality on top of core and is designed to be used after `GET /products`
and before `POST /products/{productId}/orders`. It allows users more
fine-grained control and selection of available tasking opportunities by
letting them explore the opportunities which are available for a chosen order
configuration. The opportunities are represented in a FeatureCollection, with
order specific attributes and values in the feature properties.

Implementations can choose to support synchronous or asynchronous opportunity
search requests as needed to support the duration of the operations required to
find opportunities for a given Product. In cases where the opportunity search

> can be completed within the duration of a regular HTTP request synchronous

Reviewer: ```suggestion
can be completed within the duration of a regular HTTP request, synchronous
```

operation is feasible and does not require maintaining any request or
opportunity state outside the context of the request.

> Where opportunitiy search takes a longer duration, asynchronous operation

Reviewer 2: ```suggestion
Where an opportunity search takes a longer duration, asynchronous operation
```

allows the API to return an [Opportunity Search
Record](./opportunity/README.md#opportunity-search-record) immediately, which
the client can then use to check the status of the search and retrieve the
search results once they are available.  Async support does require the
implementation to maintain the state of the search and any results outside the
context of the request, so the requestor can retrieve the results at some point
after they are made available.

## Endpoints

> STAPI follow the modern web API practices of using HTTP Request Methods

Reviewer: ```suggestion
STAPI follows the modern web API practices of using HTTP Request Methods
```

| `POST /products/{productId}/orders` | Core | [Order Request](./order/README.md#order-request) | - | Order a capture with a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters) as defined in the products or a request that was provided through the opportunities endpoint. |
| `POST /products/{productId}/opportunities` | Opportunities | [Opportunity Request](./opportunity/README.md#opportunity-request) | Sync search: [Opportunities Collection](./opportunity/README.md#opportunities-collection); Async search: [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | Explore the opportunities available for a particular set of [parameters](https://github.com/Element84/stapi-spec/blob/main/product/README.md#parameters) |
| `GET /products/{productId}/opportunities/{opportunityCollectionId}` | Opportunities (Async) | - | [Opportunities Collection](./opportunity/README.md#opportunities-collection) | Get the opportunity collection for an async opportunity search |
| `GET /searches/opportunities` | Opportunities (Async) | - | [[Opportunity Search Record](./opportunity/README.md#opportunity-search-record)] | |

> | `GET /searches/opportunities/{searchRecordId}` | Opportunities (Async) | - | [Opportunity Search Record](./opportunity/README.md#opportunity-search-record) | |

Reviewer 2: ```suggestion
| `GET /searches/opportunities/{searchRecordId}` | Opportunities (Async) | - | [Oportunity Search Record](./opportunity/README.md#oportunity-search-record) | |
```

| `GET /searches/opportunities/{searchRecordId}/statuses` | Opportunities (Async) | - | [[Opportunity Search Status](./opportunity/README.md#opportunity-search-status)] | |

> | `POST /searches/opportunities/{searchRecordId}/statuses` | Opportunities (Async) | [Opportunity Search Status](./opportunity/README.md#opportunity-search-status) | - | POST a search status update. Returns an empty 201 on success. |

Author: The whole statuses piece here borrows heavily from the `order-statuses` extension. However, the spec does not outline an equivalent of this endpoint for order statuses. I am uncertain if this is because it is intended to be an implementation-specific endpoint and not addressable by the spec or if it is an oversight.

If I should remove this endpoint from the spec let me know. Similarly, if we need to add an endpoint to update order statuses let me know that too and I can see about getting a change in for that (or at least an issue).

Author: Should be removed. `POST` endpoints are internal to the system and thus should be outside the scope of the spec, and up to the specific implementation to define.

## Conformance Classes

STAPI utilizes OGC API Features
[Conformance](http://docs.opengeospatial.org/is/17-069r3/17-069r3.html#_declaration_of_conformance_classes)
JSON structure. For STAPI, we declare new STAPI conformance classes, with the
core ones detailed in the table below.

The core STAPI conformance classes communicate the conformance JSON only in the
root (`/`) document, while OGC API requires they also live at the
`/conformance` endpoint. STAPI's conformance structure is detailed in the

> [core](core/). Note all conformance URIs serve up a rendered HTML version of
> the corresponding OpenAPI document at the given location.

Reviewer 2: ```suggestion
[core](core/). 
> [!NOTE]
> All conformance URIs serve up a rendered HTML version of
the corresponding OpenAPI document at the given location.
```

Reviewer 2: No need to implement!

| STAPI - Opportunities (Async) | [Opportunities](opportunity/README.md) | https://stapi.example.com/v0.1.0/opportunities-async | Enables async request of potential tasking opportunities |
| STAPI - Core | Core | https://geojson.org/schema/Point.json | Allows submitting orders with GeoJSON points |
| STAPI - Core | Core | https://geojson.org/schema/Linestring.json | Allows submitting orders with GeoJSON linestrings |
| STAPI - Core | Core | https://geojson.org/schema/Polygon.json | Allows submitting orders with GeoJSON polygons |
| STAPI - Core | Core | https://geojson.org/schema/MultiPoint.json | Allows submitting orders with GeoJSON multi points |
| STAPI - Core | Core | https://geojson.org/schema/MultiPolygon.json | Allows submitting orders with GeoJSON multi polygons |
| STAPI - Core | Core | https://geojson.org/schema/MultiLineString.json | Allows submitting orders with GeoJSON multi linestring |

See [the STAPI Demo](https://github.com/Element84/stat-api-demo).

## Pagination

STAPI supports paging through hypermedia links for the following resources:

- `GET /products`
- `POST /products/{productId}/opportunities`

> - `GET /products/{productId}/opportunities/{collectionId}`

Author: Good catch, it should be `/products/{productId}/opportunities/{opportunityCollectionId}`. I am updating it.

- `GET /orders`
- `GET /orders/{orderId}/statuses`
- `GET /searches/opportunities/`

> The following relation types are available for pagination:

Reviewer: ```suggestion
The following relation types may be available for pagination:
```

The href may contain any arbitrary URL parameter, which is implementation-specific:

- `https://stapi.example.com/products?page=2`

> - `https://stapi.example.com/productsnext=8a35eba9c`

Reviewer: ```suggestion
- `https://stapi.example.com/products?next=8a35eba9c`
```

    deactivate PROVIDER
```

A user with a specific product in mind views available opportunities and places

> and order.

Reviewer 2: ```suggestion
an order.
```

    PROVIDER-->>USER: Response: Order Object
    deactivate PROVIDER
```

A user with a specific product in mind makes an async opportunities search

> request, retrieves the available opportunities, and places and order.

Reviewer 2: ```suggestion
request, retrieves the available opportunities, and places an order.
```

```mermaid
sequenceDiagram
    USER->>PROVIDER: GET /products/{productId}
    activate PROVIDER
    PROVIDER-->>USER: Response: Product Object
    deactivate PROVIDER

    USER->>PROVIDER: POST /products/{productId}/opportunities
    activate PROVIDER
    PROVIDER-->>USER: Response: Opportunity Search Record
    deactivate PROVIDER

    loop Until search record is completed and has link to opportunity collection
        USER->>PROVIDER: GET /searches/opportunities/{searchRecordId}
        activate PROVIDER
        PROVIDER-->>USER: Response: Opportunity Search Record
        deactivate PROVIDER
    end

>     USER->>PROVIDER: GET /products/{productId}/opportunities/{collectionId}

Reviewer 2: ```suggestion
    USER->>PROVIDER: GET /products/{productId}/opportunities/{opportunityCollectionId}
```
